ryan_savill/principal_feature_analysis/readme.md

```diff
@@ -0,0 +1,902 @@
+# Principal Feature Analysis
+## Background
+Principal feature analysis (PFA) is a method for selecting a subset of features that describe most of the variability in the dataset based on [this paper](http://venom.cs.utsa.edu/dmz/techrep/2007/CS-TR-2007-011.pdf) by Y. Lu et al. published in ACM, 2007. Most of the code for the implementation below was provided in a [stackexchange comment section](https://stats.stackexchange.com/questions/108743/methods-in-r-or-python-to-perform-feature-selection-in-unsupervised-learning/203978#203978) and is greatly appreciated! 
+
+This sounds oddly similar to principal component analysis (PCA), which is no coincidence as the methodologies are intertwined. PCA also does a similar thing but instead of choosing features that describe the variability to a certain threshold we choose a subset of principal components. This is not always optimal and there are valid choices for working with the features instead of the principal components as the principal components are always a transformed form of the data.
+
+If you have not heard of PCA before I would advise you to check out [this video](https://www.youtube.com/watch?v=FgakZw6K1QQ). If you (like me just a few days ago) have an understanding of the principle of PCA but don't understand the math behind it I'd advise you to take some time to grasp the concepts. I found a great series of videos that explain all necessary parts of the puzzle that is PCA. Each video is just about 10 minutes long and explained quite well, even if your math lectures have been a few years ago!
+
+- [Vector Projections](https://www.youtube.com/watch?v=X78tLBY3BMk)
+- [Eigenvectors and Eigenvalues](https://www.youtube.com/watch?v=glaiP222JWA)
+- [Derivative of a Matrix](https://www.youtube.com/watch?v=e73033jZTCI)
+- [Lagrange Multipliers](https://www.youtube.com/watch?v=6oZT72-nnyI)
+- [Covariance Matrix](https://www.youtube.com/watch?v=152tSYtiQbw)
+- [PCA](https://www.youtube.com/watch?v=dhK8nbtii6I) (finally!)
+
+Once you have grasped these concepts this explanation should be a breeze!
+
+What principal feature analysis does is take advantage of some of the transformations in PCA to find out which features explain the variance. The first steps in PFA are to create the covariance matrix and find the orthonormal eigenvectors and sort them by their eigenvectors, creating a matrix A. This in effect is what PCA does as well since the principal components are the eigenvectors of the covariance matrix. We can output them in [scikit-learn](https://scikit-learn.org/stable/) with the command '.components_'. 
```

Comment:
```suggestion
Principal feature analysis takes advantage of some of the transformations in PCA to find out which features explain the variance. The first steps in PFA are to create the covariance matrix and find the orthonormal eigenvectors and sort them by their eigenvalues, creating a matrix A. This in effect is what PCA does as well since the principal components are the eigenvectors of the covariance matrix. We can output them in [scikit-learn](https://scikit-learn.org/stable/) with the command '.components_'. 
```